pkg/cloud/cloud.go

```diff
@@ -0,0 +1,52 @@
+// Copyright 2022 Google LLC All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cloud
+
+import (
+	"fmt"
+
+	agonesv1 "agones.dev/agones/pkg/apis/agones/v1"
+	"agones.dev/agones/pkg/cloud/gcp"
+	"agones.dev/agones/pkg/cloud/generic"
+	"agones.dev/agones/pkg/util/runtime"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// Cloud provides a generic interface that abstracts cloud specific functionality.
+// Users should call New() to instantiate a specific cloud provider.
+type Cloud interface {
```

Comment:
The thought I had on this the other day was that this isn't going to be used per-_Cloud_, it's going to be used per _Cloud Product_.

I.e. The aim for this is Google Cloud Autopilot, not all of Google Cloud (it won't work on GKE for example), so I'm wondering if this should be more of a `Product` flag, rather than a `Cloud` flag?

Reply:
I was planning on layering in GCP and Autopilot detection and adding basically an "is autopilot" bool to the GCP provider for this case. (As the code stands right now, though, it won't do anything unless you force `--cloud-provider=gcp` _and_ have the right Pod annotation.)

(I am imagining we may want other provider functions for GKE Standard as well.)

Reply:
I'd be tempted to make it extremely explicit - so, something like `--cloud-product=gcp-autopilot` rather than doing auto detection.

Reply:
I can make it forceable for sure, but I understood one of the goals to be portability of configuration - which would lean towards auto-detection. If we differentiate at the product level, the configuration wouldn't even be portable between GKE Standard and Autopilot, without additional work on the install side.

Reply:
We talked offline about this: @markmandel's objection was more the granularity/naming of the abstraction vs the auto-detection. I agree with him, so I changed it to a --cloud-product / CloudProduct abstraction.